transloadit: implement assembly status restore

goto-bus-stop opened the pull request:
<strike>Pretty early, mostly just opening this up so I can leave todos here.</strike>

This PR will implement restoring an upload to Transloadit after a browser crash, using GoldenRetriever. It should request new assembly statuses, set up the status WebSocket connections again, and account for any events that were missed.

Currently this happens after Golden Retriever restores an upload. I think we can use this same functionality when the internet connection died and comes `core:back-online` too. We do have information about assemblies, files, and results in state in that case, but we may have missed some events.

If we persist the entire assemblies state to localStorage for Golden Retriever, we could follow the exact same steps both for restoring from localStorage after a page refresh, and for picking things back up after a connection problem. I think.

At the moment, this:

 - [x] Fetches assembly statuses for files that were still uploading
 - [x] Reconnects to their WebSockets
 - [x] Restores assembly, files, results state
 - [x] Emits events that were missed:
    - [x] `transloadit:upload`
    - [x] `transloadit:result`
    - [x] `transloadit:assembly-error`
    - [x] `transloadit:complete`
 - [x] Does the Right Thing™ depending on `waitForEncoding` and `waitForMetadata` options
 - [x] Works when no files are being uploaded with `alwaysRunAssembly: true`
 - [x] Works when files were uploaded elsewhere first with `importFromUploadURLs`

goto-bus-stop commented:
This works now, except for retries, see #387 🙏

goto-bus-stop commented:
Two more situations to make sure this works with:

 - when no files are being uploaded with `alwaysRunAssembly: true`
 - when files were uploaded elsewhere first with `importFromUploadURLs`

There is one other problem I run into every so often now—sometimes when fetching the assembly status, it won't have any results yet, but by the time the websocket connection is set up, a result has been emitted, and by then we've missed it. That sort of thing is not likely to happen when your browser really crashed, or when an assembly takes a longer time, but with simple operations (like image resize) and accidental navigation/tab close it does … One way to tackle that could be to request the assembly status again after setting up the connection, and diffing it with the current state. Any events coming in over the connection will have already been merged into state so the diffing should work OK.

arturi commented:
Is this one ready for review? :)

goto-bus-stop commented:
yes i think it's good to go for initial review

goto-bus-stop commented:
Just rebased this on the refactored master!

goto-bus-stop commented:
The restore plugin is using a private method still to make uppy calculate progress properly after files have been restored, maybe we can make it autodetect file changes or something. I don't think the calculateProgress method should be public.

arturi commented on src/core/Core.js:
```diff
@@ -385,11 +386,37 @@ class Uppy {
+    })
+
+    this.calculateTotalProgress()
     this._calculateTotalProgress()
```
Accidental `calculateTotalProgress` here.

goto-bus-stop replied:
hello `git rebase` :rofl:  thanks!

arturi commented on src/plugins/Transloadit/index.js:
```diff
@@ -281,6 +304,191 @@ module.exports = class Transloadit extends Plugin {
+      })
+
+      this.uppy.log('[Transloadit] New fully uploaded files since restore:', newUploads)
+      newUploads.forEach(({ assembly, uploadedFile }) => {
```
The way `log()` currently works, you need to call it with the object separately for it to show it, like:

```js
this.uppy.log('[Transloadit] New fully uploaded files since restore:')
this.uppy.log(newUploads)
```

Not ideal, I know :)

goto-bus-stop replied:
oops, i just `s/console.log/this.uppy.log`d that. thanks!

arturi commented on src/plugins/Transloadit/index.js:
```diff
@@ -486,6 +707,9 @@ module.exports = class Transloadit extends Plugin {
 
+    this.uppy.on('restore:get-data', this.getPersistentData)
+    this.uppy.on('restored', this.onRestored)
+
```
So will `restore:` here a goldenretriever-specific event and `restored` a global one?

goto-bus-stop replied:
sorta. my thinking is that while `restore:get-data` is used to communicate between different plugins, Core doesn't use it. other plugins use it to give more data to GoldenRetriever when it stores restore data.

arturi commented:
> The restore plugin is using a private method still to make uppy calculate progress properly after files have been restored

Maybe add an eventlistener to Core, so it listens for `restored` and calculates progress then?

arturi commented:
Tested, works! ✨ Did bail on me a few times, but I think it was due to old localStorage stuff, was fixed when I cleaned everything.

Great job! Left some comments, feel free to merge.

goto-bus-stop commented:
> Maybe add an eventlistener to Core, so it listens for restored and calculates progress then?

oh yeah, we can just do that! thanks :D 

think i addressed all comments--will merge once travis comes back green
